File: src/states/layer4_state.py
"""Layer 4 – Pacman Hunt.

Changes from spec:
• Fog of War DISABLED – full maze always visible.
• Player starts Top-Left (0, 0).
• Both monsters start Top-Right (GRID_COLS-1, 0).
• Exit at the CENTER of the grid (GRID_COLS//2, GRID_ROWS//2).
• Two A* monsters use live pathfinding to intercept the player.
• Player picks Square or Hex grid.
"""
import random
import pygame
from src.states.base_state import BaseState
from src.maze    import generate_square_maze, generate_hex_maze, SquareGrid, HexGrid
from src.algorithms import AStarManhattan
from src.entities   import Player, Monster
from src.ui.renderer import draw_square_maze, draw_hex_maze, update_fog
from src.ui.sidebar  import draw_sidebar
from src.config import (
    GRID_COLS, GRID_ROWS,
    SIDEBAR_X, SIDEBAR_WIDTH, SCREEN_HEIGHT, MAZE_AREA_WIDTH,
    UI_BG, UI_PANEL, UI_BORDER, UI_TEXT, UI_HEADING, UI_SUBTEXT,
    UI_SUCCESS, UI_DANGER, UI_HIGHLIGHT, UI_WARNING,
    ASTAR_MON_COL, ASTAR_MAN_COL,
    font,
)

_PLAYER_START  = (0, 0)
_MON_START     = (GRID_COLS - 1, 0)                     # top-right
_EXIT          = (GRID_COLS // 2, GRID_ROWS // 2)        # centre = (5, 5)

MONSTER_INTERVAL = 0.40   # seconds between monster steps

# Two distinct monster colours
_MON_COLORS = [(180, 60, 220), (220, 100, 60)]


class Layer4State(BaseState):

    def enter(self) -> None:
        self._phase    = "choose"
        self._grid_type= None
        self._anim     = 0.0
        self._choose_rects: list = []

    def _start_game(self, grid_type: str) -> None:
        self._grid_type = grid_type
        seed = random.randint(0, 9999)

        if grid_type == "square":
            passages   = generate_square_maze(GRID_COLS, GRID_ROWS, seed)
            self._grid = SquareGrid(passages)
        else:
            passages   = generate_hex_maze(GRID_COLS, GRID_ROWS, seed)
            self._grid = HexGrid(passages)

        self.player  = Player(_PLAYER_START)

        # Both monsters spawn at top-right
        algo = AStarManhattan()
        self.monsters = [
            Monster(_MON_START, algo, _MON_COLORS[0], "A* Hunter 1"),
            Monster(_MON_START, algo, _MON_COLORS[1], "A* Hunter 2"),
        ]

        # ── No fog: reveal ALL cells immediately ──────────────────────────
        self.revealed: set = set(self._grid.passages.keys())

        self._timers = [0.0, 0.0]
        self._result = None        # None | "escaped" | "caught"
        self._result_btns: list = []
        self._phase  = "game"

        # Initial monster paths toward player start
        for m in self.monsters:
            m.compute_path(self._grid, self.player.pos)

    # ── Events ────────────────────────────────────────────────────────────────

    def handle_event(self, event: pygame.event.Event) -> None:
        if event.type == pygame.MOUSEBUTTONDOWN and event.button == 1:
            if self._phase == "choose":
                for rect, key in self._choose_rects:
                    if rect.collidepoint(event.pos):
                        self._start_game(key); return
            elif self._phase == "game" and self._result:
                for rect, target in self._result_btns:
                    if rect.collidepoint(event.pos):
                        self._go(target); return

        if event.type != pygame.KEYDOWN:
            return

        if event.key == pygame.K_ESCAPE:
            if self._phase == "game":
                self._phase = "choose"
            else:
                self._go("MENU")
            return

        if self._phase == "choose":
            if event.key == pygame.K_s:
                self._start_game("square")
            elif event.key == pygame.K_h:
                self._start_game("hex")

        elif self._phase == "game" and not self._result:
            delta = _arrow_to_delta(event.key)
            if delta:
                self._player_move(delta)

        elif self._phase == "game" and self._result:
            if event.key in (pygame.K_RETURN, pygame.K_SPACE):
                self._go("MENU")

    def _player_move(self, delta: tuple) -> None:
        c, r    = self.player.pos
        new_pos = (c + delta[0], r + delta[1])
        if not self._grid.is_passage(self.player.pos, new_pos):
            return
        self.player.move(new_pos)

        if new_pos == _EXIT:
            self._result = "escaped"
            if "MAZE_L4" not in self.game_data["completed_levels"]:
                self.game_data["completed_levels"].append("MAZE_L4")
            return

        for m in self.monsters:
            if m.pos == new_pos:
                self._result = "caught"; return

    # ── Update (real-time monster movement) ───────────────────────────────────

    def update(self, dt: float) -> None:
        self._anim += dt
        if self._phase != "game" or self._result:
            return

        for i, m in enumerate(self.monsters):
            self._timers[i] += dt
            if self._timers[i] >= MONSTER_INTERVAL:
                self._timers[i] = 0.0
                # Live A* toward player's current position
                m.compute_path(self._grid, self.player.pos)
                m.advance()
                if m.pos == self.player.pos:
                    self._result = "caught"; return

    # ── Draw ──────────────────────────────────────────────────────────────────

    def draw(self) -> None:
        self.screen.fill(UI_BG)

        if self._phase == "choose":
            self._draw_choose()
            return

        # Draw maze (no fog – all cells already in self.revealed)
        if self._grid_type == "square":
            draw_square_maze(
                self.screen, self._grid.passages, self.revealed,
                self.player.pos, _PLAYER_START, _EXIT,
                monsters=self.monsters,
            )
        else:
            draw_hex_maze(
                self.screen, self._grid.passages, self.revealed,
                self.player.pos, _PLAYER_START, _EXIT,
                monsters=self.monsters,
            )

        self._draw_sidebar()

        if self._result:
            self._draw_result()

    # ── Choose screen ─────────────────────────────────────────────────────────

    def _draw_choose(self) -> None:
        self._choose_rects = []
        s  = self.screen
        cx = MAZE_AREA_WIDTH // 2

        title = font("cn_large").render("第四关 – 追逃游戏！", True, (255,150,60))
        s.blit(title, (cx - title.get_width()//2, 60))

        lines = [
            ("Two A* monsters hunt you in real-time.", UI_TEXT),
            ("The ENTIRE maze is visible — no fog!", (80,200,255)),
            (f"Player: top-left  ·  Exit: center  ·  Monsters: top-right", UI_SUBTEXT),
            ("Reach the exit to win. Don't get caught!", UI_TEXT),
        ]
        y = 110
        for text, col in lines:
            ss = font("body").render(text, True, col)
            s.blit(ss,(cx-ss.get_width()//2,y)); y+=ss.get_height()+4

        # Grid choice buttons
        btn_data = [
            ("square", "[S]  方形迷宫", "Square Grid", ASTAR_MAN_COL),
            ("hex",    "[H]  六边形迷宫","Hex Grid",    (80,255,150)),
        ]
        bw, bh = 250, 100
        gap    = 40
        total_w= 2*bw+gap
        bx0    = (MAZE_AREA_WIDTH-total_w)//2
        by     = 200

        mx, my = pygame.mouse.get_pos()
        for i, (key, cn, en, col) in enumerate(btn_data):
            rect = pygame.Rect(bx0 + i*(bw+gap), by, bw, bh)
            hov  = rect.collidepoint(mx,my)
            pygame.draw.rect(s,(45,45,60) if hov else UI_PANEL, rect, border_radius=10)
            pygame.draw.rect(s, col, rect, width=2, border_radius=10)
            self._choose_rects.append((rect, key))
            cn_s = font("cn_body").render(cn, True, col)
            en_s = font("small").render(en, True, UI_SUBTEXT)
            s.blit(cn_s,(rect.x+(bw-cn_s.get_width())//2, rect.y+18))
            s.blit(en_s,(rect.x+(bw-en_s.get_width())//2, rect.y+54))

        # Info grid about positions
        info_y = by + bh + 30
        info_lines = [
            (f"🟢 Player spawn  : (0, 0) – top-left corner",      UI_SUCCESS),
            (f"🔴 Exit goal     : ({_EXIT[0]}, {_EXIT[1]}) – maze center",    (200,80,80)),
            (f"👾 Monster spawn : ({_MON_START[0]}, {_MON_START[1]}) – top-right corner", ASTAR_MON_COL),
            ("⚡ Monsters use real-time A* pathfinding each tick",   UI_SUBTEXT),
        ]
        for text, col in info_lines:
            ss = font("body").render(text, True, col)
            s.blit(ss,(cx-ss.get_width()//2,info_y)); info_y+=ss.get_height()+6

        draw_sidebar(
            self.screen,
            level_name   = "Layer 4 – Hunt",
            instructions = ["[S] Square","[H] Hex","ESC → Menu"],
            tomes        = self.game_data["tomes"],
        )

    # ── Race sidebar ──────────────────────────────────────────────────────────

    def _draw_sidebar(self) -> None:
        s = self.screen
        sx, sw, sh = SIDEBAR_X, SIDEBAR_WIDTH, SCREEN_HEIGHT
        pygame.draw.rect(s, UI_BG, (sx,0,sw,sh))
        pygame.draw.line(s, UI_BORDER,(sx,0),(sx,sh),2)
        x, y, w = sx+10, 12, sw-20

        # Title
        grid_cn = "方形" if self._grid_type=="square" else "六边形"
        t = font("cn_large").render(f"第四关 – {grid_cn}追逃", True, UI_HEADING)
        s.blit(t,(sx+(sw-t.get_width())//2,y)); y+=t.get_height()+6
        pygame.draw.line(s,UI_HEADING,(x,y),(x+w,y),1); y+=8

        # Status
        if self._result == "escaped":
            st_s = font("cn_body").render("✅ 成功逃出！", True, UI_SUCCESS)
        elif self._result == "caught":
            st_s = font("cn_body").render("❌ 被捉住了！", True, UI_DANGER)
        else:
            st_s = font("cn_body").render("🔴 全速逃跑！", True, UI_WARNING)
        s.blit(st_s,(x,y)); y+=st_s.get_height()+10

        # Player stats
        pygame.draw.rect(s, UI_PANEL,(x,y,w,60), border_radius=5)
        pygame.draw.rect(s, ASTAR_MAN_COL,(x,y,4,60), border_radius=2)
        s.blit(font("small").render("You",True,ASTAR_MAN_COL),(x+10,y+4))
        s.blit(font("body").render(f"Steps: {self.player.steps}",True,UI_TEXT),(x+10,y+22))
        s.blit(font("body").render(f"Time : {self.player.elapsed:.1f}s",True,UI_TEXT),(x+10,y+40))
        y+=64

        # Monster positions
        y+=6
        for i, m in enumerate(self.monsters):
            pygame.draw.rect(s, UI_PANEL,(x,y,w,44),border_radius=4)
            pygame.draw.rect(s, m.color,(x,y,4,44),border_radius=2)
            ms = font("small").render(m.name, True, m.color)
            ps = font("small").render(f"pos {m.pos}  steps {m.steps}", True, UI_TEXT)
            s.blit(ms,(x+10,y+4)); s.blit(ps,(x+10,y+24))
            y+=48

        # Info
        y+=8
        pygame.draw.line(s,UI_BORDER,(x,y),(x+w,y),1); y+=6
        for line in [
            "A* hunters recompute path",
            "to your position every tick.",
            "Full maze visible – no fog.",
            "Arrow keys to move.",
            f"Exit at centre: {_EXIT}",
        ]:
            ls = font("tiny").render(line, True, UI_SUBTEXT)
            s.blit(ls,(x,y)); y+=ls.get_height()+3

    # ── Result overlay ────────────────────────────────────────────────────────

    def _draw_result(self) -> None:
        self._result_btns = []
        s = self.screen
        overlay = pygame.Surface((MAZE_AREA_WIDTH, SCREEN_HEIGHT), pygame.SRCALPHA)
        overlay.fill((0,0,0,160)); s.blit(overlay,(0,0))
        cx, cy = MAZE_AREA_WIDTH//2, SCREEN_HEIGHT//2

        if self._result == "escaped":
            lines = [
                ("🎉 YOU ESCAPED!", UI_SUCCESS),
                ("", None),
                ("You outsmarted two real-time A* hunters!", UI_TEXT),
                ("Understanding A* helps you predict and", UI_TEXT),
                ("evade algorithmic pursuit.", UI_TEXT),
                ("", None),
                (f"Steps taken: {self.player.steps}", ASTAR_MAN_COL),
                ("", None),
            ]
        else:
            lines = [
                ("💀 CAUGHT!", UI_DANGER),
                ("", None),
                ("A* always finds the shortest path to you.", UI_TEXT),
                ("The hunters had perfect maze knowledge.", UI_TEXT),
                ("", None),
                (f"You lasted {self.player.steps} steps.", UI_SUBTEXT),
                ("", None),
            ]

        total_h = len(lines) * 28
        y = cy - total_h // 2
        for text, col in lines:
            if text and col:
                ss = font("body").render(text, True, col)
                s.blit(ss,(cx-ss.get_width()//2,y))
            y += 28

        bw, bh = 200, 42
        btn_rect = pygame.Rect(cx-bw//2, y+6, bw, bh)
        mx, my   = pygame.mouse.get_pos()
        hov      = btn_rect.collidepoint(mx,my)
        pygame.draw.rect(s, UI_PANEL if not hov else (50,50,70), btn_rect, border_radius=8)
        pygame.draw.rect(s, UI_SUCCESS, btn_rect, width=2, border_radius=8)
        ls = font("cn_body").render("← 返回主菜单", True, UI_SUCCESS)
        s.blit(ls,(btn_rect.x+(bw-ls.get_width())//2,
                   btn_rect.y+(bh-ls.get_height())//2))
        self._result_btns = [(btn_rect, "MENU")]


def _arrow_to_delta(key):
    return {
        pygame.K_UP:(0,-1), pygame.K_DOWN:(0,1),
        pygame.K_LEFT:(-1,0), pygame.K_RIGHT:(1,0),
    }.get(key)

File: web/src/states/layer4State.js
/**
 * layer4State.js — Layer 4: The Hunt
 * Real-time: two A* monsters chase the player across a fully-visible maze.
 * Player starts top-left (0,0), monsters at top-right (9,0), exit at centre (5,5).
 * Mirrors src/states/layer4_state.py
 */

import { posKey, keyToPos }                            from '../utils.js';
import { GRID_COLS, GRID_ROWS, HEX_NBR_EVEN, HEX_NBR_ODD } from '../config.js';
import { generateSquareMaze }                           from '../maze/generator.js';
import { generateHexMaze }                              from '../maze/generator.js';
import { SquareGrid }                                   from '../maze/squareGrid.js';
import { HexGrid }                                      from '../maze/hexGrid.js';
import { AStarManhattan }                               from '../algorithms/astar.js';
import {
    SCREEN_HEIGHT, MAZE_AREA_WIDTH,
    SIDEBAR_X, SIDEBAR_WIDTH, C,
    drawSquareMaze, drawHexMaze, drawSidebarBase,
    drawText, drawRoundRect, drawStatBlock,
} from '../renderer.js';

// ── Fixed positions ───────────────────────────────────────────────────────────
const PLAYER_START_K = posKey(0, 0);
const MON_START_K    = posKey(GRID_COLS - 1, 0);
const EXIT_K         = posKey(Math.floor(GRID_COLS / 2), Math.floor(GRID_ROWS / 2));

const MONSTER_INTERVAL = 0.40;   // seconds between each monster step

const MON_COLORS = ['#b43cdc', '#dc643c'];   // purple, orange
const MON_NAMES  = ['A* Hunter 1', 'A* Hunter 2'];

// ── Key mappings ──────────────────────────────────────────────────────────────
const SQ_MOVE_KEYS = {
    ArrowUp:[0,-1], ArrowDown:[0,1], ArrowLeft:[-1,0], ArrowRight:[1,0],
    w:[0,-1], s:[0,1], a:[-1,0], d:[1,0],
};

// Hex direction indices: 0=NE 1=E 2=SE 3=SW 4=W 5=NW
const HEX_KEY_DIR = { i:0, k:1, m:2, n:3, h:4, u:5 };

// ── Entity factories ──────────────────────────────────────────────────────────

function makePlayer(startK) {
    return {
        posK:      startK,
        steps:     0,
        startTime: performance.now(),
        move(k)  { this.posK = k; this.steps++; },
        get elapsed() { return (performance.now() - this.startTime) / 1000; },
    };
}

function makeMonster(startK, color, name) {
    return {
        posK:  startK,
        color, name,
        path:    [],
        pathIdx: 0,
        steps:   0,

        computePath(grid, targetK) {
            const { path } = new AStarManhattan().findPath(grid, this.posK, targetK);
            this.path    = path;
            this.pathIdx = 0;
        },

        advance() {
            if (this.pathIdx + 1 >= this.path.length) return;
            this.pathIdx++;
            this.posK = this.path[this.pathIdx];
            this.steps++;
        },
    };
}

// ── Factory ───────────────────────────────────────────────────────────────────

export function createLayer4State(ctx, goTo, gameData) {
    let phase;          // 'choose' | 'game'
    let gridType;       // 'square' | 'hex'
    let grid, player, monsters, revealed, timers, result;
    let anim = 0;
    let mouseX = 0, mouseY = 0;
    let chooseBtnRects = [];   // [{x,y,w,h,key}, ...]
    let resultBtnRect  = null;

    // ── Lifecycle ─────────────────────────────────────────────────────────────

    function enter() {
        phase  = 'choose';
        anim   = 0;
        result = null;
        chooseBtnRects = [];
        resultBtnRect  = null;
    }

    function _startGame(type) {
        gridType = type;
        const seed = Math.floor(Math.random() * 10000);

        if (type === 'square') {
            const passages = generateSquareMaze(GRID_COLS, GRID_ROWS, seed);
            grid = new SquareGrid(passages);
        } else {
            const passages = generateHexMaze(GRID_COLS, GRID_ROWS, seed);
            grid = new HexGrid(passages);
        }

        player   = makePlayer(PLAYER_START_K);
        monsters = MON_COLORS.map((col, i) => makeMonster(MON_START_K, col, MON_NAMES[i]));
        timers   = [0, 0];
        result   = null;
        chooseBtnRects = [];
        resultBtnRect  = null;

        // No fog — reveal every cell immediately
        revealed = new Set(grid.passages.keys());

        phase = 'game';

        // Pre-compute initial paths toward player
        for (const m of monsters) m.computePath(grid, player.posK);
    }

    // ── Input ─────────────────────────────────────────────────────────────────

    function handleKey(key) {
        if (key === 'Escape') {
            if (phase === 'game') { phase = 'choose'; return; }
            goTo('MENU'); return;
        }

        if (phase === 'choose') {
            if (key === 's' || key === 'S') { _startGame('square'); return; }
            if (key === 'h' || key === 'H') { _startGame('hex');    return; }
            return;
        }

        if (phase === 'game' && result) {
            if (key === 'Enter' || key === ' ') { goTo('MENU'); }
            return;
        }

        if (phase === 'game' && !result) {
            if (gridType === 'square') {
                const delta = SQ_MOVE_KEYS[key];
                if (!delta) return;
                const [c, r] = keyToPos(player.posK);
                const newK   = posKey(c + delta[0], r + delta[1]);
                _playerMove(newK);
            } else {
                const dir = HEX_KEY_DIR[key];
                if (dir === undefined) return;
                const [c, r]   = keyToPos(player.posK);
                const nbrs     = (r % 2 === 0) ? HEX_NBR_EVEN : HEX_NBR_ODD;
                const [dc, dr] = nbrs[dir];
                const newK     = posKey(c + dc, r + dr);
                _playerMove(newK);
            }
        }
    }

    function _playerMove(newK) {
        if (!grid.isPassage(player.posK, newK)) return;
        player.move(newK);

        if (newK === EXIT_K) {
            result = 'escaped';
            if (!gameData.completedLevels.includes('MAZE_L4'))
                gameData.completedLevels.push('MAZE_L4');
            return;
        }

        for (const m of monsters) {
            if (m.posK === newK) { result = 'caught'; return; }
        }
    }

    function handleMouseMove(x, y) { mouseX = x; mouseY = y; }

    function handleClick(x, y) {
        if (phase === 'choose') {
            for (const btn of chooseBtnRects) {
                if (_hit(btn, x, y)) { _startGame(btn.key); return; }
            }
        }
        if (phase === 'game' && result && resultBtnRect && _hit(resultBtnRect, x, y)) {
            goTo('MENU');
        }
    }

    // ── Update (real-time monsters) ───────────────────────────────────────────

    function update(dt) {
        anim += dt;
        if (phase !== 'game' || result) return;

        for (let i = 0; i < monsters.length; i++) {
            timers[i] += dt;
            if (timers[i] >= MONSTER_INTERVAL) {
                timers[i] = 0;
                const m = monsters[i];
                m.computePath(grid, player.posK);   // live re-target
                m.advance();
                if (m.posK === player.posK) { result = 'caught'; return; }
            }
        }
    }

    // ── Draw ──────────────────────────────────────────────────────────────────

    function draw() {
        if (phase === 'choose') {
            _drawChoose(); return;
        }

        // Draw maze — all cells revealed, no fog
        const monOpts = { monsters: monsters.map(m => ({ posK: m.posK, color: m.color })) };
        if (gridType === 'square') {
            drawSquareMaze(ctx, grid.passages, revealed, player.posK,
                           PLAYER_START_K, EXIT_K, monOpts);
        } else {
            drawHexMaze(ctx, grid.passages, revealed, player.posK,
                        PLAYER_START_K, EXIT_K, monOpts);
        }

        _drawSidebar();
        if (result) _drawResultOverlay();
    }

    // ── Choose screen ─────────────────────────────────────────────────────────

    function _drawChoose() {
        // Background
        ctx.fillStyle = C.uiBg;
        ctx.fillRect(0, 0, MAZE_AREA_WIDTH, SCREEN_HEIGHT);

        const cx = MAZE_AREA_WIDTH / 2;

        // Title
        drawText(ctx, 'Layer 4 – The Hunt', cx, 50,
                 { size: 28, bold: true, color: '#ff963c', align: 'center' });
        drawText(ctx, 'Two A* monsters hunt you in real-time.', cx, 92,
                 { size: 15, color: C.uiText, align: 'center' });
        drawText(ctx, 'Full maze visible — no fog!', cx, 114,
                 { size: 14, color: '#50c8ff', align: 'center' });
        drawText(ctx, `Player: (0,0) top-left  ·  Exit: (${Math.floor(GRID_COLS/2)},${Math.floor(GRID_ROWS/2)}) center  ·  Monsters: (${GRID_COLS-1},0) top-right`,
                 cx, 136, { size: 12, color: C.uiSubtext, align: 'center' });

        // Grid type buttons
        const btnData = [
            { key: 'square', label: 'Square Grid', sublabel: '[S] to select', color: C.astarMan },
            { key: 'hex',    label: 'Hex Grid',    sublabel: '[H] to select', color: '#50ff96' },
        ];

        const bw = 220, bh = 110, gap = 50;
        const totalW = bw * 2 + gap;
        const bx0 = (MAZE_AREA_WIDTH - totalW) / 2;
        const by  = 168;

        chooseBtnRects = [];
        for (let i = 0; i < btnData.length; i++) {
            const { key, label, sublabel, color } = btnData[i];
            const bx  = bx0 + i * (bw + gap);
            const btn = { x: bx, y: by, w: bw, h: bh, key };
            chooseBtnRects.push(btn);

            const hov = _hit(btn, mouseX, mouseY);
            drawRoundRect(ctx, bx, by, bw, bh, 10,
                          { fill: hov ? '#1e1e36' : C.uiPanel, stroke: color, lineWidth: 2 });
            drawText(ctx, label,    bx + bw/2, by + 28, { size: 18, bold: true, color, align: 'center' });
            drawText(ctx, sublabel, bx + bw/2, by + 60, { size: 13, color: C.uiSubtext, align: 'center' });
        }

        // Info block
        const infoY = by + bh + 28;
        const infoLines = [
            { text: `Player spawns at (0,0) — top-left corner`,            color: C.uiSuccess },
            { text: `Exit at (${Math.floor(GRID_COLS/2)},${Math.floor(GRID_ROWS/2)}) — maze center`,  color: '#e05050' },
            { text: `Two monsters spawn at (${GRID_COLS-1},0) — top-right`, color: MON_COLORS[0] },
            { text: 'Monsters recompute A* path to YOU every 0.4s',         color: C.uiSubtext },
            { text: 'Reach the exit before being caught!',                   color: C.uiText },
        ];
        let iy = infoY;
        for (const { text, color } of infoLines) {
            drawText(ctx, text, cx, iy, { size: 13, color, align: 'center' });
            iy += 24;
        }

        // Square controls hint
        drawText(ctx, 'Square: Arrow keys / WASD   ·   Hex: I U K H M N', cx, iy + 14,
                 { size: 12, color: C.uiSubtext, align: 'center' });

        // Sidebar
        _drawChooseSidebar();
    }

    function _drawChooseSidebar() {
        drawSidebarBase(ctx);
        const sx = SIDEBAR_X + 10, sw = SIDEBAR_WIDTH - 20;
        let y = 14;

        ctx.font = 'bold 20px Arial'; ctx.fillStyle = C.uiHeading;
        ctx.textAlign = 'center'; ctx.textBaseline = 'top';
        ctx.fillText('Layer 4 – The Hunt', SIDEBAR_X + SIDEBAR_WIDTH / 2, y); y += 32;

        _ln(sx, y, sx + sw, y); y += 12;

        drawText(ctx, 'Choose a grid type to begin.', sx + sw/2, y,
                 { size: 13, color: C.uiSubtext, align: 'center' }); y += 32;

        _ln(sx, y, sx + sw, y); y += 12;

        const rules = [
            'Two A* hunters chase you live.',
            'Full maze visible — no fog.',
            'Reach center exit to escape.',
            "Don't step on a monster!",
            '[S] Square   [H] Hex',
            'ESC → Main Menu',
        ];
        for (const line of rules) {
            drawText(ctx, line, sx, y, { size: 12, color: C.uiSubtext }); y += 20;
        }
    }

    // ── Race sidebar ──────────────────────────────────────────────────────────

    function _drawSidebar() {
        drawSidebarBase(ctx);
        const sx = SIDEBAR_X + 10, sw = SIDEBAR_WIDTH - 20;
        let y = 14;

        const gridLabel = gridType === 'square' ? 'Square' : 'Hex';
        ctx.font = 'bold 19px Arial'; ctx.fillStyle = C.uiHeading;
        ctx.textAlign = 'center'; ctx.textBaseline = 'top';
        ctx.fillText(`Layer 4 – ${gridLabel} Hunt`, SIDEBAR_X + SIDEBAR_WIDTH / 2, y); y += 28;
        _ln(sx, y, sx + sw, y); y += 10;

        // Status line
        let statusText, statusColor;
        if (result === 'escaped') {
            statusText = 'YOU ESCAPED!'; statusColor = C.uiSuccess;
        } else if (result === 'caught') {
            statusText = 'CAUGHT!'; statusColor = C.uiDanger;
        } else {
            statusText = 'RUN!'; statusColor = C.uiWarn;
        }
        drawText(ctx, statusText, sx + sw/2, y,
                 { size: 16, bold: true, color: statusColor, align: 'center' }); y += 28;
        _ln(sx, y, sx + sw, y); y += 10;

        // Player stat block
        y = drawStatBlock(ctx, sx, y, sw,
            'You  (player)', player.steps, player.elapsed, C.player) + 10;

        // Monster stat blocks
        for (const m of monsters) {
            drawRoundRect(ctx, sx, y, sw, 54, 5, { fill: C.uiPanel });
            ctx.fillStyle = m.color;
            ctx.beginPath(); ctx.roundRect(sx, y, 4, 54, 2); ctx.fill();
            drawText(ctx, m.name,         sx + 10, y + 4,  { size: 12, bold: true, color: m.color });
            drawText(ctx, `pos  (${keyToPos(m.posK).join(', ')})`, sx + 10, y + 22, { size: 12, color: C.uiText });
            drawText(ctx, `steps  ${m.steps}`,                     sx + 10, y + 38, { size: 12, color: C.uiSubtext });
            y += 58;
        }
        y += 6;

        _ln(sx, y, sx + sw, y); y += 8;

        const notes = [
            'A* hunters recompute path',
            'to your position every 0.4s.',
            'Full maze visible – no fog.',
            gridType === 'square'
                ? 'Move: Arrows / WASD'
                : 'Move: I U K H M N',
            `Exit at centre: (${Math.floor(GRID_COLS/2)},${Math.floor(GRID_ROWS/2)})`,
            'ESC → back to grid choice',
        ];
        for (const line of notes) {
            drawText(ctx, line, sx, y, { size: 12, color: C.uiSubtext }); y += 18;
        }
    }

    // ── Result overlay ────────────────────────────────────────────────────────

    function _drawResultOverlay() {
        ctx.fillStyle = 'rgba(0,0,0,0.72)';
        ctx.fillRect(0, 0, MAZE_AREA_WIDTH, SCREEN_HEIGHT);

        const pw = 540, ph = 360;
        const px = (MAZE_AREA_WIDTH - pw) / 2;
        const py = (SCREEN_HEIGHT   - ph) / 2;

        const escaped      = result === 'escaped';
        const accentColor  = escaped ? C.uiSuccess : C.uiDanger;

        drawRoundRect(ctx, px, py, pw, ph, 14,
                      { fill: '#14142a', stroke: accentColor, lineWidth: 2 });

        // Headline
        const headline = escaped ? 'YOU ESCAPED!' : 'CAUGHT!';
        drawText(ctx, headline, px + pw/2, py + 20,
                 { size: 26, bold: true, color: accentColor, align: 'center' });

        ctx.strokeStyle = C.uiBorder; ctx.lineWidth = 1;
        _ln(px + 20, py + 60, px + pw - 20, py + 60);

        let iy = py + 74;
        if (escaped) {
            const lines = [
                { text: 'You outsmarted two real-time A* hunters!', color: C.uiText },
                { text: 'Understanding A* helps you predict',        color: C.uiSubtext },
                { text: 'and evade algorithmic pursuit.',            color: C.uiSubtext },
            ];
            for (const { text, color } of lines) {
                drawText(ctx, text, px + pw/2, iy, { size: 14, color, align: 'center' }); iy += 26;
            }
            iy += 6;
            drawText(ctx, `Steps taken: ${player.steps}`, px + pw/2, iy,
                     { size: 20, bold: true, color: C.astarMan, align: 'center' }); iy += 34;
            drawText(ctx, `Time: ${player.elapsed.toFixed(1)}s`, px + pw/2, iy,
                     { size: 16, color: C.uiSubtext, align: 'center' }); iy += 30;
        } else {
            const lines = [
                { text: 'A* always finds the shortest path to you.', color: C.uiText },
                { text: 'The hunters had perfect maze knowledge.',    color: C.uiText },
            ];
            for (const { text, color } of lines) {
                drawText(ctx, text, px + pw/2, iy, { size: 14, color, align: 'center' }); iy += 26;
            }
            iy += 10;
            drawText(ctx, `You lasted ${player.steps} steps.`, px + pw/2, iy,
                     { size: 16, color: C.uiSubtext, align: 'center' }); iy += 34;
        }

        // Pedagogical insight
        _ln(px + 20, iy, px + pw - 20, iy); iy += 14;
        const insight = [
            { text: 'Why is A* so effective at chasing?', color: C.uiHeading },
            { text: 'h=|Δx|+|Δy| guides monsters directly toward you,', color: C.uiSubtext },
            { text: 'expanding far fewer nodes than Dijkstra would.', color: C.uiSubtext },
        ];
        for (const { text, color } of insight) {
            drawText(ctx, text, px + pw/2, iy, { size: 13, color, align: 'center' }); iy += 22;
        }

        iy += 6;

        // Button
        const btnW = 200, btnH = 42;
        const bx = px + (pw - btnW) / 2;
        const by = py + ph - 56;
        resultBtnRect = { x: bx, y: by, w: btnW, h: btnH };
        const hov = _hit(resultBtnRect, mouseX, mouseY);
        drawRoundRect(ctx, bx, by, btnW, btnH, 8, {
            fill: hov ? '#1e2e1e' : C.uiPanel,
            stroke: C.uiSuccess, lineWidth: 2,
        });
        drawText(ctx, '← Back to Menu', bx + btnW/2, by + btnH/2 - 9,
                 { size: 15, bold: true, color: C.uiSuccess, align: 'center' });
        drawText(ctx, 'Enter or click', bx + btnW/2, by + btnH - 16,
                 { size: 11, color: C.uiSubtext, align: 'center' });
    }

    // ── Helpers ───────────────────────────────────────────────────────────────

    function _ln(x1, y1, x2, y2) {
        ctx.beginPath(); ctx.moveTo(x1, y1); ctx.lineTo(x2, y2); ctx.stroke();
    }

    return { enter, update, draw, handleKey, handleClick, handleMouseMove };
}

function _hit(r, x, y) {
    return x >= r.x && x <= r.x + r.w && y >= r.y && y <= r.y + r.h;
}
